### PR TITLE
Feat/amp grad state

### LINKED ISSUE #67590: Add `OptState.UPDATED` to `torch.cuda.amp.GradScaler`
## 🚀 Feature

Add `OptState.UPDATED` to `torch.cuda.amp.GradScaler` to signal that the scaler has been stepped and `optimizer.step()` has been actually called.

## Motivation

`GradScaler` skips the `optimizer.step()` call when Inf/NaNs are found. This means that users can accidentally assume that `optimizer.step()` has been called if they don't know their model produced these types of values.

An example of a bug induced by this is https://github.com/pytorch/pytorch/issues/44511 where we want to run `scheduler.step()` only after `optimizer.step()`, but there's no easy way to be sure that happened when the `GradScaler` is used.

Having this would also make testing the scaler integration simpler for ecosystem libraries.

## Pitch

```diff
def step(self, optimizer, *args, **kwargs):
    if (not self._enabled):
        return optimizer.step(*args, **kwargs)

    if "closure" in kwargs:
        raise RuntimeError("Closure use is not currently supported if GradScaler is enabled.")

    self._check_scale_growth_tracker("step")

    optimizer_state = self._per_optimizer_states[id(optimizer)]

    if optimizer_state["stage"] is OptState.STEPPED:
        raise RuntimeError("step() has already been called since the last update().")

    retval = None

    if (hasattr(optimizer, "_step_supports_amp_scaling") and optimizer._step_supports_amp_scaling):
        # This optimizer has customized scale-handling logic, so we can call optimizer.step() directly.
        # The contract with custom optimizers is that their step() should accept an additional,
        # optional grad_scaler kwarg.  We append self to the kwargs so the custom optimizer has full information:
        # it can query its own state, invoke unscale_ on itself, etc
        retval = optimizer.step(*args, **dict(kwargs, grad_scaler=self))       
-       optimizer_state["stage"] = OptState.STEPPED
+       optimizer_state["stage"] = OptState.UPDATED
        return retval

    if optimizer_state["stage"] is OptState.READY:
        self.unscale_(optimizer)

    assert len(optimizer_state["found_inf_per_device"]) > 0, "No inf checks were recorded for this optimizer."

    retval = self._maybe_opt_step(optimizer, optimizer_state, *args, **kwargs)

-   optimizer_state["stage"] = OptState.STEPPED
-
    return retval


def _maybe_opt_step(self, optimizer, optimizer_state, *args, **kwargs):
-   retval = None
    if not sum(v.item() for v in optimizer_state["found_inf_per_device"].values()):
        retval = optimizer.step(*args, **kwargs)
+       optimizer_state["stage"] = OptState.UPDATED
+       return retval
+   optimizer_state["stage"] = OptState.STEPPED
+   return
```

This feature request is particularly valuable with the implementation of #67589

## Alternatives

Not do it. Or have users subclass `GradScaler`.

## Additional context

The warning `UserWarning: Detected call of lr_scheduler.step() before optimizer.step()` is impacting Lightning (https://github.com/PyTorchLightning/pytorch-lightning/issues/5558) because of this silent skipping behaviour and we haven't found a clean way to fix it. cc @akihironitta

### PR DESCRIPTION
**Pull request by GdoongMathew**
Fixes #67590 

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -5976,6 +5957,45 @@ def run(model0, model1, optimizer0, optimizer1, try_scaling_api):
...
+
+            opt_state = scaler._per_optimizer_states[id(opt)]
+            opt_state["found_inf_per_device"] = scaler._unscale_grads_(
+                opt,
+                inv_scale=torch.tensor([1.0], device=device),
+                found_inf=found_inf,
+                allow_fp16=False,
+            )
+
+            scaler._maybe_opt_step(opt, opt_state)
```
Comment (jgong5) [inline test/test_torch.py:5992]: Can this issue only be exposed and tested by checking internal methods?

Comment (GdoongMathew) [reply test/test_torch.py:5992]: Hi @jgong5, thanks for reviewing.

So the main goal of this feature request is to provide a robust way for users to distinguish different stages of the optimizer, so that they can decide whether to call `lr_scheduler.step` later on - particularly in cases where the optimizer is skipped internally by the `GardScaler`. 

To reliably trigger this condition (i.e. producing `inf` or `nan` gradient), I used part of the logic in another test:
https://github.com/pytorch/pytorch/blob/20398498ead9419f88d183ebac37a61fc19face0/test/test_torch.py#L5536-L5541

this allows us to deterministically simulate the scenario where the optimizer step is skipped.

Comment (GdoongMathew) [reply test/test_torch.py:5992]: Hi @jgong5, a friendly bump.
